Backports for julia v1.10.0-beta2

KristofferC opened the pull request:
Backported PRs:
- [x] #50637 <!-- Remove SparseArrays legacy code -->
- [x] #50665 <!-- print `@time` msg into print buffer -->
- [x] #50523 <!-- Avoid generic call in most cases for getproperty -->
- [x] #50635 <!-- `versioninfo()`: include build info and unofficial warning -->
- [x] #50670 <!-- Make reinterpret specialize fully. -->
- [x] #50666 <!-- include `--pkgimage=no` caches for stdlibs -->
- [x] #50765 
- [x] #50764
- [x] #50768
- [x] #50767
- [x] #50618 <!-- inference: continue const-prop' when concrete-eval returns non-inlineable -->
- [x] #50689 <!-- Attach `tanpi` docstring to method -->
- [x] #50671 <!-- Fix rdiv of complex lhs by real factorizations -->
- [x] #50598 <!-- only limit types in stack traces in the REPL -->
- [x] #50766 <!-- Don't partition alwaysinline functions -->
- [x] #50771 <!-- re-allow non-string values in ENV `get!` -->
- [x] #50682 <!-- Add fallback if we have make a weird GC decision. -->
- [x] #50781 <!-- fix `bit_map!` with aliasing -->
- [x] #50172 <!-- print feature flags used for matching pkgimage -->
- [x] #50844 <!-- Bump OpenBLAS binaries to use the new GEMM multithreading threshold -->
- [x] #50826 <!-- Update dependency builds -->
- [x] #50845 <!-- fix #50438, use default pool for at-threads -->
- [x] #50568 <!-- `Array(::AbstractRange)` should return an `Array` -->
- [x] #50655 <!-- fix hashing regression. -->
- [x] #50779 <!-- Minor refactor to image generation -->
- [x] #50791 <!-- Make symbols internal in jl_create_native, and only externalize them when partitioning -->
- [x] #50724 <!-- Merge opaque closure modules with the rest of the workqueue -->
- [x] #50738 <!-- Add alignment to constant globals -->
- [x] #50871 <!-- macOS: Don't inspect dead threadtls during exception handling. -->

Need manual backport:

Contains multiple commits, manual intervention needed:

Non-merged PRs with backport label:
- [ ] #50850 <!-- Remove weird Rational dispatch and add pi functions to list -->
- [ ] #50823 <!-- Make ranges more robust with unsigned indexes. -->
- [ ] #50809 <!-- Limit type-printing in MethodError -->
- [ ] #50663 <!-- Fix Expr(:loopinfo) codegen -->
- [ ] #50594 <!-- Disallow non-index Integer types in isassigned -->
- [ ] #50385 <!-- Precompile pidlocks: add to NEWS and docs -->
- [ ] #49805 <!-- Limit TimeType subtraction to AbstractDateTime -->

jishnub committed:
`Array(::AbstractRange)` should return an `Array` (#50568)

Currently, `Array(r::AbstractRange)` falls back to `vcat(r)`, but
certain ranges may choose to specialize `vcat(r::AbstractRange)` to not
return an `Array`. This PR ensures that `Array(r)` always returns an
`Array`.

At present, there's some code overlap with `vcat` (just above the
`Array` method added in this PR). Perhaps some of these may be replaced
by `unsafe_copyto!`, but the tests for ranges include some special cases
that don't support `getindex`, which complicates things a bit. I've not
done this for now. In any case, the common bit of code is pretty simple,
so perhaps the duplication is harmless.

(cherry picked from commit 3cc05908a61aec885f3a4fc3fc9cb13cf3524108)

dkarrasch committed:
Remove SparseArrays legacy code (#50637)

(cherry picked from commit c57d33a17ae713497052cddbba83a5f4b5aaeaa3)

IanButterworth committed:
print `@time` msg into print buffer (#50665)

(cherry picked from commit 912423456dfe9f54c7454efd252a496aa1ea579f)

gbaraldi committed:
Avoid generic call in most cases for getproperty (#50523)

More generic than https://github.com/JuliaLang/julia/pull/50444.

Should we keep the boundscheck on the getfield since we got the index
from the runtime itself?

(cherry picked from commit 207c09a6b1e22fe6148de664f1e718c06faf12fb)

oscardssmith committed:
fix hashing regression. (#50655)

This fixes 2 bugs introduced by #49996 and #50041.
Closes #50628.

(cherry picked from commit c777c714ff03c76ebbda07992639e8bced0983e2)

pchintalapudi committed:
Reuse incremental JIT compilation for --image-codegen (#50649)

We don't need to merge all of the workqueue modules when performing
compilation with `--image-codegen` set, we just need the global variable
initializers to be defined before they're used in one of the modules.
Therefore we can do this by compiling all of the global variable
initializers upfront, so that later references will link them properly.

(cherry picked from commit ff14eafd803558f4c1f9ecdeecc44d3a4caa7343)

IanButterworth committed:
`versioninfo()`: include build info and unofficial warning (#50635)

(cherry picked from commit c43e5a10be27b7f93b5368875aa1d2596b4d4947)

gbaraldi committed:
Make reinterpret specialize fully. (#50670)

Fixes https://github.com/JuliaLang/julia/issues/50612

The issue here was the reinterpret change made a bunch of operations
like `Core.bitcast(UInt64,24)` not fold, even though they are fully
known at compile time. That made `UInt32(Char)` not inline which then
caused the regression.

(cherry picked from commit dc064683087d017927381e64c38bd162aea47b4e)

IanButterworth committed:
include `--pkgimage=no` caches for stdlibs (#50666)

(cherry picked from commit 62605cc40f51f5a921e5a9bdced1821afd49183f)

sostock committed:
Attach `tanpi` docstring to method (#50689)

(cherry picked from commit 5c69561b0028ae97fc120b6a4714bfc2eb11a947)

dkarrasch committed:
Fix rdiv of complex lhs by real factorizations (#50671)

Co-authored-by: Martin Holters <martin@holters.name>
(cherry picked from commit 210c5b526b828a2335232300252b2f113b41ccb5)

pchintalapudi committed:
Merge opaque closure modules with the rest of the workqueue (#50724)

This sticks the compiled opaque closure module into the
`compiled_functions` list of modules that we have compiled for the
particular `jl_codegen_params_t`. We probably should manage that vector
in codegen_params, since it lets us see if a particular codeinst has
already been compiled but not yet emitted.

(cherry picked from commit 441fcb1e06513ceff4d5ca4e111863c2ebba957b)

JeffBezanson committed:
only limit types in stack traces in the REPL (#50598)

fixes #50575

Co-authored-by: Shuhei Kadowaki <40514306+aviatesk@users.noreply.github.com>
(cherry picked from commit 98222572628fa67d733389cc718f2c70d45d2ea1)

pchintalapudi committed:
Add alignment to constant globals (#50738)

(cherry picked from commit 37082297452099ec262cd6100fe2aaa1717d7aa9)

pchintalapudi committed:
Don't partition alwaysinline functions (#50766)

(cherry picked from commit bea8c444141c9cbbf4a3f6515ce572a28be3c0a1)

aviatesk committed:
effects: fix `:nothrow` modeling of `getglobal` (#50765)

aviatesk committed:
irinterp: refine `:nothrow` only when it is not proved yet (#50764)

This makes irinterp not override `:nothrow=true` that are assumed by
`Base.@assume_effects`.

aviatesk committed:
ssair: `compact!` constant `PiNode` (#50768)

Currently the `compact!`-ion pass fails to fold constant `PiNode` like
`PiNode(0.0, Const(0.0))`. This commit fixes it up.

aviatesk committed:
optimizer: move inlineable constants into argument position during `compact!`-ion (#50767)

In code like below
```julia
Base.@assume_effects :nothrow function erase_before_inlining(x, y)
    z = sin(y)
    if x
        return "julia"
    end
    return z
end

let y::Float64
    length(erase_before_inlining(true, y))
end
```
the constant prop' can figure out the constant return type of
`erase_before_inlining(true, y)` while it is profitable not to inline
expand it since otherwise we left some `!:nothrow` callees there
(xref: https://github.com/JuliaLang/julia/pull/47305).

In order to workaround this problem, this commit makes `compact!`move
inlineable constants into argument positions so that the such
"inlineable, but safe as a whole" calls to be erased during compaction.
This should give us general compile-time performance improvement too
as we no longer need to expand the IR for those calls.

Requires:
- #50764
- #50765
- #50768

JeffBezanson committed:
re-allow non-string values in ENV `get!` (#50771)

fixes #50472

(cherry picked from commit 6dd763bf2a0b130f03183e99c0ebf0a3c14c97d6)

pchintalapudi committed:
Minor refactor to image generation (#50779)

(cherry picked from commit f337c3d261c8618396506f3c56f5bde3479f66bf)

gbaraldi committed:
Add fallback if we have make a weird GC decision. (#50682)

If something odd happens during GC (the PC goes to sleep) or a very big
transient the heuristics might make a bad decision. What this PR
implements is if we try to make our target more than double the one we
had before we fallback to a more conservative method. This fixes the new
issue @vtjnash found in https://github.com/JuliaLang/julia/issues/40644
for me.

(cherry picked from commit ab94fadc9f8cd83c4c1550ac43483d4e0b1a3e9c)

aviatesk committed:
inference: continue const-prop' when concrete-eval returns non-inlineable (#50618)

oscardssmith committed:
fix `bit_map!` with aliasing (#50781)

fixes #50780 caused by #47013.

(cherry picked from commit 3e04129d61e19fe2957680b39e03b350db8e8c0d)

vchuravy committed:
print feature flags used for matching pkgimage (#50172)

```
julia> @ccall jl_dump_host_cpu()::Cvoid
CPU: znver2
Features: sse3, pclmul, ssse3, fma, cx16, sse4.1, sse4.2, movbe, popcnt, aes, xsave, avx, f16c, rdrnd, fsgsbase, bmi, avx2, bmi2, rdseed, adx, clflushopt, clwb, sha, rdpid, sahf, lzcnt, sse4a, prfchw, mwaitx, xsaveopt, xsavec, xsaves, clzero, wbnoinvd

julia> target = only(Base.current_image_targets())
znver2; flags=0; features_en=(sse3, pclmul, ssse3, fma, cx16, sse4.1, sse4.2, movbe, popcnt, aes, xsave, avx, f16c, fsgsbase, bmi, avx2, bmi2, adx, clflushopt, clwb, sha, rdpid, sahf, lzcnt, sse4a, prfchw, mwaitx, xsavec, xsaves, clzero, wbnoinvd)
```

Co-authored-by: Prem Chintalapudi <prem.chintalapudi@gmail.com>
Co-authored-by: Jameson Nash <vtjnash@gmail.com>
(cherry picked from commit 958da95647da5f9a824f99cb56846eed22a4d098)

pchintalapudi committed:
Make symbols internal in jl_create_native, and only externalize them when partitioning (#50791)

(cherry picked from commit 8b8da91ad7701c5401947a749399720f5aa51418)

ViralBShah committed:
Bump OpenBLAS binaries to use the new GEMM multithreading threshold (#50844)

Detailed discussion and benchmarks by @oscardssmith in
https://github.com/JuliaPackaging/Yggdrasil/pull/7189

(cherry picked from commit 626f6872ed8043b1565623a4653bc50f7e9a07cc)

ararslan committed:
Update dependency builds (#50826)

This bumps the build numbers for stdlib and binary dependency JLLs,
updates libssh2 to 1.11.0, libgit2 to 1.6.4, and objconv to 2.53.

Julia's FreeBSD CI has been running on FreeBSD 13.2 for a while, but
until more recently, Yggdrasil was still building FreeBSD binaries using
the 12.2 sysroot. The sysroot was updated to 13.2 and I went through and
rebuilt the dependencies that Julia uses. The updated build numbers
correspond to these rebuilt but otherwise unchanged binaries.

The actual version updates are because libssh2 in Yggdrasil was at
1.11.0 so I left it there (its [release
notes](https://github.com/libssh2/libssh2/releases/tag/libssh2-1.11.0)
suggest it's a safe update), libgit2 had a newer patch version available
and needed to be fixed anyway since the Windows build was broken, and
objconv needed its Yggdrasil build recipe fixed but Elliot's GitHub
mirror of objconv was at 2.53 so I updated to use that.

(cherry picked from commit 5e51fbe78df4304d7f4ca776928144905ba494bf)

JeffBezanson committed:
fix #50438, use default pool for at-threads (#50845)

(cherry picked from commit 744aa79c2c2655de526f02dcd1987ae1d7db6f6d)

SBuercklin committed:
Add `Base.get_extension` to docs/API (#50860)

The documentation of `Pkg.jl` is instructing
developers to rely on the existence of this function, and there doesn't
seem to be any alternative which is a part of the API that developers
can use instead on to guarantee forward compatibility for the same
behavior.

`Base.get_extension` is [referred to
explicitly](https://pkgdocs.julialang.org/v1.9/creating-packages/#Backwards-compatibility)
in the `Pkg.jl` docs to conditionally use package extensions vs
`Requires.jl`.

The `Pkg.jl` docs suggest
```julia
if !isdefined(Base, :get_extension)
  include("../ext/PlottingContourExt.jl")
end
```
to transition from "normal dependency to extension," which will break
and automatically load the extension in future versions should
`Base.get_extension` go away.

`Base.get_extension` is the only way (that I know of) to directly access
the module associated with a package extension, which can be a useful
utility as well.

(cherry picked from commit d1759bcde8e662e6bb7aaff94466b3ca1f3cf43b)

KristofferC commented:
@nanosoldier `runtests(; vs=":release-1.9")`

KristofferC commented:
@nanosoldier `runtests(ALL, vs = ":release-1.9")`

andrebsguedes committed:
Makes IntrusiveLinkedListSynchronized mutable to avoid allocation on poptask (#50802)

Currently `poptask` is allocating every time it calls into
`jl_task_get_next` due to `StickyWorkqueue`
(`IntrusiveLinkedListSynchronized`) being immutable and requiring an
allocation to be used as `Any` on the `ccall`.

The allocations can be seen in the following snippet:
```
function work()
    done = 0.0
    l = Threads.SpinLock()
    Threads.@sync for _ in 1:(Threads.nthreads() / 2)
        Threads.@spawn begin
            v = 1.0
            for i in 1:(1<<33)
                v *= 1.0001
                if i % (1 << 13) == 0
                    yield()
                end
            end
            Base.@lock_nofail l done += v
        end
    end
    return done
end

julia> @time work()
 15.758794 seconds (5.03 M allocations: 153.523 MiB, 0.35% gc time)
```

Which after the change becomes:
```
julia> @time work()
 15.907513 seconds (67 allocations: 4.719 KiB)
```

(cherry picked from commit 2f03072436e2a40549e35b56a84c40b71abbdf5a)

pchintalapudi committed:
Combined compilation fixes

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50708#issuecomment-1672852498) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/27bc777_vs_6fc1be0/report.html) is available.

maleadt committed:
macOS: Don't inspect dead threadtls during exception handling. (#50871)

Otherwise we may end up with the state of a dead thread, as the system
IDs can alias. This can lead to an early return from the exception
handler, resulting in e.g. safepoint exceptions being actually delivered
to user code.

---------

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
Co-authored-by: Cody Tapscott <topolarity@tapscott.me>

gbaraldi committed:
Change addprocs limit to total_memory instead of free memory.

(cherry picked from commit 5e3fb41e349f0a50d6ecd8315717b6234c1b5fbb)
